src/client/dfs/duns.c

```diff
@@ -184,60 +176,90 @@ duns_link_path(const char *path, const char *sysname,
 			      attrp->da_chunk_size);
 		if (len < 90) {
 			D_ERROR("Failed to create xattr value\n");
-			D_GOTO(err_pool, rc = -DER_INVAL);
+			D_GOTO(err_link, rc = -DER_INVAL);
 		}
 
 		rc = lsetxattr(path, DUNS_XATTR_NAME, str, len + 1, 0);
 		if (rc) {
 			D_ERROR("Failed to set DAOS xattr (rc = %d).\n", rc);
-			D_GOTO(err_pool, rc = -DER_INVAL);
+			D_GOTO(err_link, rc = -DER_INVAL);
 		}
 
-		rc = daos_cont_create(poh, attrp->da_cuuid, NULL, NULL);
+		if (attrp->da_type == DAOS_PROP_CO_LAYOUT_POSIX) {
+			dfs_attr_t	dfs_attr;
+
+			/** TODO: set Lustre FID here. */
+			dfs_attr.da_id = 0;
+			dfs_attr.da_oclass_id = attrp->da_oclass_id;
+			dfs_attr.da_chunk_size = attrp->da_chunk_size;
+			rc = dfs_cont_create(poh, attrp->da_cuuid, &dfs_attr,
+					     NULL, NULL);
+		} else {
+			daos_prop_t	*prop;
+
+			prop = daos_prop_alloc(1);
+			if (prop == NULL) {
+				D_ERROR("Failed to allocate container prop.");
+				D_GOTO(err_link, rc = -DER_NOMEM);
+			}
+			prop->dpp_entries[0].dpe_type =
+				DAOS_PROP_CO_LAYOUT_TYPE;
+			prop->dpp_entries[0].dpe_val = attrp->da_type;
+			rc = daos_cont_create(poh, attrp->da_cuuid, prop, NULL);
+			daos_prop_free(prop);
+		}
 	} while ((rc == -DER_EXIST) && try_multiple);
 	if (rc) {
 		D_ERROR("Failed to create container (%d)\n", rc);
-		D_GOTO(err_pool, rc);
+		D_GOTO(err_link, rc);
 	}
 
-	/*
-	 * TODO: Add a container attribute to store the inode number (or Lustre
-	 * FID) of the file.
-	 */
+	return rc;
+err_link:
+	if (attrp->da_type == DAOS_PROP_CO_LAYOUT_HDF5)
+		unlink(path);
+	else if (attrp->da_type == DAOS_PROP_CO_LAYOUT_POSIX)
+		rmdir(path);
+	return rc;
+}
+
+int
+duns_destroy_path(daos_handle_t poh, const char *path)
+{
+	struct duns_attr_t dattr = {0};
+	int	rc;
 
-	/** If this is a POSIX container, create DFS mount */
-	if (attrp->da_type == DAOS_PROP_CO_LAYOUT_POSIX) {
-		daos_handle_t		coh;
-		daos_cont_info_t	co_info;
-		dfs_t			*dfs;
+	/* Resolve pool, container UUIDs from path */
+	rc = duns_resolve_path(path, &dattr);
+	if (rc) {
+		D_ERROR("duns_resolve_path() Failed on path %s (%d)\n",
+			path, rc);
+		return rc;
+	}
 
-		rc = daos_cont_open(poh, attrp->da_cuuid, DAOS_COO_RW, &coh,
-				    &co_info, NULL);
+	if (dattr.da_type == DAOS_PROP_CO_LAYOUT_HDF5) {
+		rc = unlink(path);
 		if (rc) {
-			D_ERROR("Failed to open container (%d)\n", rc);
-			D_GOTO(err_cont, rc = 1);
+			D_ERROR("Failed to unlink file %s: %s\n",
+				path, strerror(errno));
+			return -DER_INVAL;
 		}
-
-		rc = dfs_mount(poh, coh, O_RDWR, &dfs);
-		dfs_umount(dfs);
-		daos_cont_close(coh, NULL);
+	} else if (dattr.da_type == DAOS_PROP_CO_LAYOUT_POSIX) {
+		rc = rmdir(path);
 		if (rc) {
-			D_ERROR("dfs_mount failed (%d)\n", rc);
-			D_GOTO(err_cont, rc = 1);
+			D_ERROR("Failed to remove dir %s: %s\n",
+				path, strerror(errno));
+			return -DER_INVAL;
 		}
 	}
 
-	daos_pool_disconnect(poh, NULL);
-	return rc;
+	/** Destroy the container */
+	rc = daos_cont_destroy(poh, dattr.da_cuuid, 1, NULL);
```

Comment:
I think the container should be destroyed first and then the path to be removed, this to prefer dangling links to leaked containers ...

Reply:
agreed